crates/driver/src/driver.rs

```diff
@@ -34,14 +35,46 @@ impl Driver {
         &self,
         auction: Auction,
     ) -> Result<SettlementSummary, SolveError> {
-        let fetch_liquidity_from_block = block_number(&self.block_stream.borrow())?;
+        self.solve_until_deadline(auction)
+            .await
+            .map_err(SolveError::from)
+    }
+
+    /// Computes a solution with the liquidity collected from a given block.
+    async fn compute_solution_for_block(
+        &self,
+        auction: Auction,
+        block: Block,
+    ) -> Result<SettlementSummary> {
+        let block = block_number(&block)?;
         let auction = self
             .auction_converter
-            .convert_auction(auction, fetch_liquidity_from_block)
+            .convert_auction(auction, block)
             .await?;
-        let summary = self.solver.commit(auction).await?;
-        tracing::info!(?summary, "computed winning settlement summary");
-        Ok(summary)
+        self.solver.commit(auction).await
+    }
+
+    /// Keeps solving the given auction with updated liquidity on every new block or until the
+    /// auction deadline is reached.
+    async fn solve_until_deadline(&self, auction: Auction) -> Result<SettlementSummary> {
+        let compute_solutions = into_stream(self.block_stream.clone())
+            .then(|block| self.compute_solution_for_block(auction.clone(), block));
+        // TODO get deadline from autopilot auction
+        let timeout = tokio::time::sleep(tokio::time::Duration::from_secs(25));
+        tokio::pin!(timeout, compute_solutions);
+
+        let mut current_solution = Err(anyhow::anyhow!("reached the deadline without a result"));
+        loop {
+            tokio::select! {
+                new_solution = compute_solutions.next() => {
```

Comment:
It was a hard fight with the type system to address this issue but it should be resolved now.
The code you [suggested](https://github.com/cowprotocol/services/pull/475#pullrequestreview-1089948607) was a good starting point but the thing ended up a bit more complicated than expected because I needed to forward the block info produced by the block stream to the function which solves the auction on that block.

Reply:
Judging by the second [example](https://docs.rs/tokio-stream/latest/tokio_stream/wrappers/struct.WatchStream.html) my very simple original solution would have been alright. 😭 
If I understood the docs correctly the stream would not have produced `N+1` at all because it wouldn't have been the latest value at the time of polling (after finishing the computation).

Reply:
😭

Reply:
I double checked and, yes you are right, `WatchStream` implements exactly what we want.

Reply:
I ran the tests against my original solution (with a minor adjustment) and they all passed. Although I feel a bit stupid to revert to my original solution I think it's probably best to have the reduced complexity. 😓